modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java

```diff
@@ -337,50 +337,63 @@ void sendUndeployRequest(String rsrcName, Collection<ClusterNode> rmtNodes) thro
                 GridIoPolicy.P2P_POOL);
         }
     }
-
+    
+    /**
+     * Queries whether a node is a direct or indirectly responsible for 
+     * the request being processed by this thread.
+     * 
+     * @param  node The node in question
+     * @return true if the node is active
+     */
+    boolean nodeOriginatedCurrentRequest(final ClusterNode node) {
+       // activeReqNodeIds is thread local.
+       Collection<UUID> nodeIds = activeReqNodeIds.get();
+       
+       return (nodeIds != null && nodeIds.contains(node.id()));
+    }
+    
     /**
      * Sends request to the remote node and wait for response. If there is
-     * no response until threshold time, method returns null.
+     * no response until threshold time, method returns null.  The receiver
+     * may forward the request to other nodes, but not to nodes that have been
+     * searched already, or nodes we intend to search directly.
      *
+     * The exclusion list severely weakens the potential n-squared behavior when a 
+     * resource does not exist. If the resource does not exist, all nodes 
+     * in the graph must be touched at least once.   To avoid all multiple touches 
+     * for the resource-not-found case, the response would need to include the 
+     * nodes forwarded to, so they could be added to the exclusion list.  However
+     * that seems like overkill for this kinds of graphs peer class loading will generate.
```

Comment:
that seems like overkill for the kinds of graphs peer class loading will generate.